### ISSUE TITLE
when I execute the project DigitRecognizerCNN I get an error

### ISSUE THREAD

**Issue by unnamed082** [mlpack/examples#35]
win8.1 x64
qt creator 5.12.0 , mingw 7.3.0

I copied the mlpack / models folder and indicated the paths to the models / Kaggle / kaggle_utils.hpp file
the program breaks inside the for loop (int i = 0; i <= CYCLES; i ++)

text error:
Сигнатура проблемы:
  Имя события проблемы:	APPCRASH
  Имя приложения:	mlpack.exe
  Версия приложения:	0.0.0.0
  Отметка времени приложения:	5dee23f7
  Имя модуля с ошибкой:	msvcrt.dll
  Версия модуля с ошибкой:	7.0.9600.17415
  Отметка времени модуля с ошибкой:	545055fe
  Код исключения:	c0000005
  Смещение исключения:	000000000000188f
  Версия ОС:	6.3.9600.2.0.0.768.101
  Код языка:	1049
  Дополнительные сведения 1:	200a
  Дополнительные сведения 2:	200a18060e7de70916b37b7f9ae679ec
  Дополнительные сведения 3:	645c
  Дополнительные сведения 4:	645c95ed87e559442dba34b2b8b13721

please tell me how to fix it?

**Comment by zoq** [comment]
Hello @unnamed082, can you make sure dataset is loaded correctly, e.g. what is the output of:

```c++
std::cout << tempDataset.n_rows << std::endl;
std::cout << tempDataset.n_cols << std:.endl;
```

**Comment by unnamed082** [comment]
Hello @zoq 
785
42001
if i check after load tempDataser

and i find that program crashes on this row in cycle:
model.Train(trainX, trainY, optimizer);

**Comment by zoq** [comment]
Sorry for the slow response, I'm wondering if this error is specific to the system you are running on or the example, can you check if the test case:

https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/convolutional_network_test.cpp#L32

fails as well?

**Comment by unnamed082** [comment]
this example runs without errors

**Comment by unnamed082** [comment]
constexpr int CYCLES = 1;

      constexpr double STEP_SIZE = 10e-3;//5e-3;

      constexpr int BATCH_SIZE = 1;

      cout << "Reading data ..." << endl;


      arma::Mat<unsigned char> matrix;
      data::ImageInfo info;
      data::Load("../berlin_0.png", matrix, info, false, true);


      size_t samples = 1;
      const arma::mat trainX(matrix.n_rows, samples);//преобразование из Mat<unsigned char> в mat(double)
                                                     //для функции train

      cout << info.Width() << " " << info.Height() << " " << info.Channels();


      const int ITERATIONS_PER_CYCLE = 1;//trainX.n_cols;

      FFN<NegativeLogLikelihood<>, RandomInitialization> model;


      model.Add<Convolution<> >(
            3,  // Number of input activation maps.
            6,  // Number of output activation maps.
            5,  // Filter width.
            5,  // Filter height.
            1,  // Stride along width.
            1,  // Stride along height.
            0,  // Padding width.
            0,  // Padding height.
            info.Width(), // Input width.
            info.Height()  // Input height.
            );

      model.Add<ReLULayer<> >();

      model.Add<MaxPooling<> >(
          2,  // Width of field.
          2,  // Height of field.
          2,  // Stride along width.
          2,  // Stride along height.
          true
          );

      model.Add<Linear<> >(6*2510*2398, 2);
      model.Add<LogSoftMax<> >();

      cout << "Training ..." << endl;

      // Setting parameters Stochastic Gradient Descent (SGD) optimizer.
      SGD<AdamUpdate> optimizer(
        STEP_SIZE,
        BATCH_SIZE,
        ITERATIONS_PER_CYCLE,
        1e-8,
        true,
        AdamUpdate(1e-8, 0.9, 0.999));

      mat trainY;
      trainY << 1;


      cout << trainY.n_rows << " " << trainY.n_cols;

      for (int i = 1; i <= CYCLES; i++)
      {

        model.Train(trainX, trainY, optimizer);

      }


if I try to train this example, I get such an error
image size 5024 * 4800 * 3 channels

**Comment by OO7kartik** [comment]
could you share the command you used to compile?
and how did you give the path to models / Kaggle / kaggle_utils.hpp?

**Comment by adithya-tp** [comment]
Hey there @unnamed082 ! I am trying to look into this issue as well, but i need to fix a problem with loading images, before i can get to the issue at hand.

The code that i used is a follows (exactly the same as yours with slight chages such as the imports):


      #include <mlpack/core.hpp>
      #include <mlpack/core/data/split_data.hpp>
      #include <mlpack/methods/ann/layer/layer.hpp>
      #include <mlpack/methods/ann/ffn.hpp>
      #include <ensmallen.hpp>
      #define STB_IMAGE_IMPLEMENTATION
      #include <mlpack/core/data/load_image_impl.hpp>

      using namespace mlpack;
      using namespace mlpack::ann;
      using namespace arma;
      using namespace std;
      using namespace ens;

      int main()
      {
          constexpr double RATIO = 0.1;
          constexpr int CYCLES = 1;
          const int ITERATIONS_PER_CYCLE = 1;
          constexpr double STEP_SIZE = 10e-3;
          constexpr int BATCH_SIZE = 1;

          cout << "Reading data ..." << endl;
          arma::Mat<unsigned char> matrix;
          data::ImageInfo info;
          data::Load("saitama.png", matrix, info, false, true);

          size_t samples = 1;
          const arma::mat trainX(matrix.n_rows, samples);

          cout << info.Width() << " " << info.Height() << " " << info.Channels();

          FFN<NegativeLogLikelihood<>, RandomInitialization> model;

          model.Add<Convolution<> >(
                3,  // Number of input activation maps.
                6,  // Number of output activation maps.
                5,  // Filter width.
                5,  // Filter height.
                1,  // Stride along width.
                1,  // Stride along height.
                0,  // Padding width.
                0,  // Padding height.
                info.Width(), // Input width.
                info.Height()  // Input height.
                );

          model.Add<ReLULayer<> >();

          model.Add<MaxPooling<> >(
              2,  // Width of field.
              2,  // Height of field.
              2,  // Stride along width.
              2,  // Stride along height.
              true
              );

          model.Add<Linear<> >(6*2510*2398, 2);
          model.Add<LogSoftMax<> >();

          cout << "Training ..." << endl;

          // Setting parameters Stochastic Gradient Descent (SGD) optimizer.
          SGD<AdamUpdate> optimizer(
            STEP_SIZE,
            BATCH_SIZE,
            ITERATIONS_PER_CYCLE,
            1e-8,
            true,
            AdamUpdate(1e-8, 0.9, 0.999));

          arma::mat trainY;
          trainY << 1;
          cout << trainY.n_rows << " " << trainY.n_cols;

          for (int i = 1; i <= CYCLES; i++)
          {
             model.Train(trainX, trainY, optimizer);
          }
          return 0;
      }
              
However i keep getting the following error:

`terminate called after throwing an instance of 'std::runtime_error'
  what():  Load(): HAS_STB is not defined, so STB is not available and images cannot be loaded!
Aborted (core dumped)`

Would love to know what I'm doing wrong here. @unnamed082 or @zoq , if you could help me on this issue that would be great. The image itself doesn't seem to be the issue, but that's just what I think. Please feel free to correct me on this. Personally I feel there is something wrong with my import statements, but I'm not sure.

I understand mlpack's ability to load images is a new addition, which is why I guess I coudn't find any documentation on how to use it. 

That being said, I would love to add this documentation myself to the "Load" docs, if anyone could help with this problem. It would be a great asset for new-comers trying to figure their way around mlpack. Thanks!

**Comment by unnamed082** [comment]
You have problems connecting and using the STB library, I encountered such a problem when I created 2 projects with the same name and there were internal Cmake errors. Try creating an empty project and plugging in the STB library and then into your project.

**Comment by adithya-tp** [comment]
@unnamed082 , shouldn't the number of output nodes in the "Linear Layer" = 1, since trainY has only one value?
that is, shouldn't it be 
`model.Add<Linear<> >(6*2510*2398, 1);` 
and not 
`model.Add<Linear<> >(6*2510*2398, 2);` ?

**Comment by adithya-tp** [comment]
@unnamed082 , oh and the HAS_STB error was resolved by adding the macro while compiling the file, namely, i used, 
```g++ cnn.cpp -o cnn `pkg-config --cflags --libs mlpack armadillo` -fopenmp -DHAS_STB```

**Comment by unnamed082** [comment]
I think in the output layer we indicate the number of classes
trainY has a column / row equal to 1, but the values inside are different
I can’t look at an example now, but I remember that
if I am not mistaken

**Comment by adithya-tp** [comment]
Okay. The following is related to your issue, but not exactly. However, the code is quite similar so I decided to post it here. When I try to train the following cnn on a single picture, it only runs when I include the line 

`arma::conv_to<arma::mat>::from(matrix).print("Print Image: \n");`

However, if I comment out this line, the code no longer runs (both cases compile. but in this case doesn't run). It doesn't make sense to me why this happens. Here's the code: 

    #include <mlpack/core.hpp>
    #include <mlpack/core/data/split_data.hpp>
    #include <mlpack/methods/ann/layer/layer.hpp>
    #include <mlpack/methods/ann/ffn.hpp>
    #include <ensmallen.hpp>

    using namespace mlpack;
    using namespace mlpack::ann;
    using namespace arma;
    using namespace std;
    using namespace ens;

    int main()
    {
        double RATIO = 0.1;
        int CYCLES = 10;
        const int ITERATIONS_PER_CYCLE = 1;
        double STEP_SIZE = 10e-3;
        int BATCH_SIZE = 1;

        // Reading in image.
        cout << "Reading data ..." << endl;
        arma::Mat<unsigned char> matrix;
        data::ImageInfo info;
        data::Load("saitama.png", matrix, info, false, true);

        FFN<NegativeLogLikelihood<>, RandomInitialization> model;

        model.Add<Convolution<> >(info.Channels(), 6, 5, 5, 1, 1, 0, 0, info.Width(), info.Height());
        model.Add<ReLULayer<> >();
        model.Add<MaxPooling<> >(2, 2, 2, 2, true);
        model.Add<Linear<> >(62*62*6, 2);
        model.Add<LogSoftMax<> >();
        cout << "Training ..." << endl;

        // Setting parameters Stochastic Gradient Descent (SGD) optimizer.
        SGD<AdamUpdate> optimizer(
          STEP_SIZE,
          BATCH_SIZE,
          ITERATIONS_PER_CYCLE,
          1e-8,
          true,
          AdamUpdate(1e-8, 0.9, 0.999));

        arma::mat trainY;
        trainY << 1;// << 0;
        arma::conv_to<arma::mat>::from(matrix).print("Print Image: \n");
        trainY.print("Train label: \n");
        cout << trainY.n_rows << " " << trainY.n_cols;

        for (int i = 1; i <= CYCLES; i++)
        {
          model.Train(arma::conv_to<arma::mat>::from(matrix), trainY, optimizer, PrintLoss(), ProgressBar());
        }
        return 0;
    }

(The image size I used is 128 x 128).
I think I should open up a separate issue about this. However, I would love to get anyone's feedback on this issue. Thanks!
